x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java

```diff
@@ -289,6 +289,7 @@ private static LifecycleExecutionState updateExecutionStateToStep(
 
         // clear any step info or error-related settings from the current step
         updatedState.setFailedStep(null);
+        updatedState.setPreviousStepInfo(existingState.stepInfo());
```

Comment:
Don't we also want these statements here https://github.com/elastic/elasticsearch/blob/b977a337e3d28e67e37e5d36634a81a89ef9e705/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java#L183-L183
and here?
https://github.com/elastic/elasticsearch/blob/b977a337e3d28e67e37e5d36634a81a89ef9e705/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/IndexLifecycleTransition.java#L429

Or am I missing a reason why we don't need them there?

Reply:
I don't believe so... Line 183 is in a method that calls `updateExecutionStateToStep(...)` on line 173 so the copy should already have happened.

Line 429 is in the method that writes the step info into the internal cluster state so any modifications of the step info will have happened further up the call chain already.

Reply:
The line numbers you quoted don't seem to line up with what I'm seeing in IndexLifecycleTransition.java :-/ Line 101 is the end of a throw statement and 102 is just a closing bracket... Could you confirm the file and commit hash those line numbers refer to? 

But talking in general about when the update cluster state method is called, I guess there may be a difference in how I interpreted the requirements... I targeted this to only copy the `step_info` at the start of a step (IE to preserve it across runs) and for that copy of the previous attempt to remain unchanged through the next run attempt.

If we were to add copy logic to the method at 429 (Or methods calling it), `previous_step_info` would get updated and overwritten each time the step progress is written to `step_info` during the execution of a step, meaning error messages from a previous run might disappear as that attempt progresses, not really fixing the underlying SDH request for it to stick around during the whole next attempt.

It's also entirely possible I haven't understood things correctly 😅 Feel free to grab me for a quick face to face chat if you think I'm not grasping the logic here :-)

Reply:
It looks like Mary is referring to:
https://github.com/elastic/elasticsearch/blob/80dd56398f855923df0c0117abd098fd3023ce36/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/ExecuteStepsUpdateTask.java#L101-L101

I think your point about not wanting to override `previous_step_info` on _every_ `step_info` makes sense. Otherwise, in the scenario that the issue/SDH describes, `previous_step_info` would get overridden on the second automatic retry attempt. So line 429 indeed doesn't need to be updated - and 183 is indeed already covered by the `updateExecutionStateToStep` call before.

Reply:
I see, so we only set the previous info when we update the cluster state with the next step not during it's execution. Clear thanks!